client/authority-discovery/src/lib.rs

```diff
@@ -211,19 +211,39 @@ where
 			.encode(&mut serialized_addresses)
 			.map_err(Error::EncodingProto)?;
 
-		for key in self.get_priv_keys_within_authority_set()?.into_iter() {
-			let signature = key.sign(&serialized_addresses);
+		let keys: Vec<CryptoTypePublicPair> = self.get_own_public_keys_within_authority_set()?
+			.into_iter()
+			.map(Into::into)
+			.collect();
 
+		let signatures = self.key_store
+			.read()
+			.sign_with_all(
+				key_types::AUTHORITY_DISCOVERY,
+				keys.clone(),
+				serialized_addresses.as_slice(),
+			)
+			.map_err(|_| Error::SigningFailed)?;
+
+		for (index, (sign_result, key)) in signatures.iter().zip(keys).enumerate() {
 			let mut signed_addresses = vec![];
+
+			// sign_with_all returns Option<Signature> where the signature
+			// is None for a public key that is not supported.
+			// Verify that all signatures exist for all provided keys.
+			if sign_result.is_err() {
```

Comment:
I would prefer something like the following to not have an `unwrap`:

```suggestion
			let signature = sign_result.map_err(|_| Err(Error::MissingSignature(index)))?;
```